File: app/src/main/java/wangdaye/com/geometricweather/db/dao/HistoryDao.java
package wangdaye.com.geometricweather.db.dao;

import androidx.annotation.Nullable;
import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.Query;

import java.util.Date;
import java.util.List;

import wangdaye.com.geometricweather.db.entities.HistoryEntity;

@Dao
public interface HistoryDao {

    @Insert
    void insertHistoryEntity(HistoryEntity entity);

    @Query("DELETE FROM HISTORY_ENTITY WHERE CITY_ID = :cityId AND WEATHER_SOURCE = :source")
    void deleteLocationHistoryEntity(String cityId, String source);

    @Query("SELECT * FROM HISTORY_ENTITY WHERE CITY_ID = :cityId AND WEATHER_SOURCE = :source "
            + "AND DATE >= :from AND DATE < :to LIMIT 1")
    @Nullable
    HistoryEntity selectHistoryEntity(String cityId, String source, Date from, Date to);

    @Query("SELECT * FROM HISTORY_ENTITY WHERE CITY_ID = :cityId AND WEATHER_SOURCE = :source")
    List<HistoryEntity> selectHistoryEntityList(String cityId, String source);
}

File: data/src/main/java/wangdaye/com/geometricweather/db/dao/HistoryDao.kt
package wangdaye.com.geometricweather.db.dao

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.Query
import java.util.Date
import wangdaye.com.geometricweather.db.entities.HistoryEntity

@Dao
interface HistoryDao {

    @Insert
    fun insertHistoryEntity(entity: HistoryEntity)

    @Query("DELETE FROM HISTORY_ENTITY WHERE CITY_ID = :cityId AND WEATHER_SOURCE = :source")
    fun deleteLocationHistoryEntity(cityId: String, source: String)

    @Query(
        "SELECT * FROM HISTORY_ENTITY WHERE CITY_ID = :cityId AND WEATHER_SOURCE = :source " +
            "AND DATE >= :from AND DATE < :to LIMIT 1"
    )
    fun selectHistoryEntity(cityId: String, source: String, from: Date, to: Date): HistoryEntity?

    @Query("SELECT * FROM HISTORY_ENTITY WHERE CITY_ID = :cityId AND WEATHER_SOURCE = :source")
    fun selectHistoryEntityList(cityId: String, source: String): List<HistoryEntity>
}
